### ISSUE TITLE
Scripted updating of individual WP-CLI packages

### ISSUE THREAD

**Issue by danielbachhuber** [wp-cli/wp-cli#3924]
When we need to apply some change across all packages, it would be nice to do so in a scripted manner.

For instance, we need to apply https://github.com/wp-cli/scaffold-package-command/pull/89 across https://github.com/wp-cli/media-command, https://github.com/wp-cli/package-command, etc.

Related https://github.com/wp-cli/wp-cli/issues/3857

**Comment by schlessera** [comment]
That would be very useful, although we'd need to do some heavy thinking about how to do this in a safe and reliable way. First thoughts are that we would need to have a list of meta data about the individual packages, that we could then run through templates.

**Comment by schlessera** [comment]
I thought some more about this. To move quickly, I suggest the following:

To run commands across packages, I suggest using [`gr`](https://github.com/mixu/gr) for now. You can set it up in a matter of minutes, and it allows you to run arbitrary commands across a subset of the repos (like `gr @wp-cli-commands composer update`).

To build scripts to simplify common tasks, I suggest using [`robo`](http://robo.li/). It allows us to build arbitrary functionality, has built-in support for VCS, Phars, etc... It will be one file at the root of the organization folder (provided your folder layout mirrors the GitHub org setup).

I also suggest building a generalized behat feature that tests for all that we want to see in an external command. Stuff like having the correct version strings in the requirements, proper Travis setup, etc... Every time we make a change to our "ideal setup", we should not only change the scaffolding system, but also the tests described above to make sure we are actually using the latest changes across all packages as well. Travis will then help us keep all of our packages at the same state.

**Comment by danielbachhuber** [comment]
Would you be opposed to using WP-CLI for this?

**Comment by schlessera** [comment]
The problem with WP-CLI is that it needs a WordPress install to even start doing something.

**Comment by schlessera** [comment]
I actually don't really care about what tool to use, `gr` just happens to be ready and you can set it up in 10 minutes' time. We should at least consider it as a temporary solution.

**Comment by danielbachhuber** [comment]
> The problem with WP-CLI is that it needs a WordPress install to even start doing something.

Not necessarily. Use the `@when before_wp_load` decoration to run the command before WordPress loads (skipping the check for WordPress entirely)

**Comment by danielbachhuber** [comment]
> We should at least consider it as a temporary solution.

1. Temporary solutions have a tendency to become permanent solutions.
2. We're not in any rush for a solution, so better to do it right the first time.

**Comment by schlessera** [comment]
> Temporary solutions have a tendency to become permanent solutions.

Oh, yes, I can affirm... :)

I wonder, though, where do we draw the line between "makes sense to include in WP-CLI" and "I have a hammer, everything looks like a nail"?

Running some kind of upgrade procedure across a multitude of GitHub repositories... does that still have anything to do with the "command line interface to WordPress"? It is on the command line, yes, but it does not have anything to do with WordPress for my understanding.

So, more generally, are we building WP-CLI-CLI? :)

**Comment by danielbachhuber** [comment]
I understand the concern.

For the relative simplicity of what we need to do to solve this problem (iterate through directories and perform a one-off modification), I think it's easier to do this within our existing toolset (WP-CLI commands) or bash, than it is to introduce a new tool, dependency, complexity, etc.

A WP-CLI command gives us a bit more power than a bash script, but I'm not opposed to using the latter.

**Comment by danielbachhuber** [comment]
Here's my bash script to indicate bundled commands as bundled:

```
#!/bin/bash

set -ex

cd $1
git checkout master
git checkout -b clarify-bundled
jq --indent 4 '.extra.bundled = true' composer.json | sponge composer.json
composer dump-autoload && wp scaffold package-readme ./ --force
git commit . -m 'Clarify this command as a bundled command'

MESSAGE="Clarify this command as a bundled command

See https://github.com/wp-cli/wp-cli/issues/4046"

git push origin clarify-bundled
hub pull-request -m "$MESSAGE"
```

**Comment by danielbachhuber** [comment]
@schlessera `grun` is pretty nice :)

**Comment by danielbachhuber** [comment]
```
ls -a vendor/wp-cli/ | grep command | xargs grun +@wp-cli-packages
```

**Comment by swissspidy** [comment]
https://github.com/wp-cli/wp-cli-dev/blob/4e4021245cf041d2c88e43832eef746f8ab8184d/foreach-bundle
